VirtAddr improvements

josephlr opened the pull request:
The first commit removes our `*const T -> VirtAddr` conversion on 32-bit platforms. Per the comment, it's "only here for backwards compatibility". Should we go even further and remove the conversions on all non-`x86_64` platforms?

The second commit makes it so `new` and `try_new` fail on non-canonical addresses. Fixes #299 

The third commit make `new` and `try_new` const. Other `VirtAddr`/`PhysAddr` methods are made const in #369 (which will be pulled back into this branch when its merged).

phil-opp commented on src/addr.rs:
```diff
@@ -66,27 +66,29 @@ impl VirtAddr {
-            0 | 0x1ffff => Ok(VirtAddr(addr)),     // address is canonical
-            1 => Ok(VirtAddr::new_truncate(addr)), // address needs sign extension
-            _ => Err(VirtAddrNotValid(addr)),
+    /// This function tries to performs sign extension of bit 47 to make the
+    /// address canonical. It succeeds if bits 48 to 64 are a correct sign
+    /// extension (i.e. copies of bit 47). Else, an error is returned.
```
These docs are a bit confusing. The function only does a sign extension to check if the address is canonical, so I wouldn't mention that implementation detail in the docs. Maybe something like this:

```suggestion
    /// This function checks wether the given address is canonical
    /// and returns an error otherwise. An address is canoncial
    /// if bits 48 to 64 are a correct sign
    /// extension (i.e. copies of bit 47).
```

phil-opp reviewed (approved):
Looks good to me overall!

We should also update the docs of `new_truncate` because it still mentions `new` and `try_new` as alternative methods with similar behavior, but these methods don't do any sign extension anymore. So maybe remove that last sentence, or change it to "If you want to check whether an address is canonical ...".

josephlr commented:
> We should also update the docs of `new_truncate` because it still mentions `new` and `try_new` as alternative methods with similar behavior, but these methods don't do any sign extension anymore. So maybe remove that last sentence, or change it to "If you want to check whether an address is canonical ...".

I updated the docs for `new` and `try_new` to make it clear that they don't sign extend anything (they just check if addresses are canonical). I also made your recommended changes to `new_truncate`, added some cross links, and fixed some typos.
